assessments-service/src/main/resources/db/migration/common/V1.43__update_indexes.sql

```diff
@@ -0,0 +1,4 @@
+ALTER TABLE `Assessment` ADD INDEX `assessment_programme_membership_id` (`programmeMembershipId`);
+ALTER TABLE `Assessment` ADD INDEX `assessment_review_date` (`reviewDate`);
+ALTER TABLE `Assessment` ADD INDEX `assessment_soft_deleted_date` (`softDeletedDate`);
+ALTER TABLE `AssessmentOutcome` ADD INDEX `assessment_outcome_outcome` (`outcome`);
```

Comment:
Assessment searches use additional unindexed fields which we might want to index for performance reasons, namely the 6 fields mentioned here: https://github.com/Health-Education-England/TIS-ASSESSMENTS/blob/a12067c1fc089b90af90defb355b8b1c85387afa/assessments-service/src/main/java/com/transformuk/hee/tis/assessment/service/service/impl/AssessmentServiceImpl.java#L147

Reply:
My expectation is that these indices would be less effective than an index combining softDeletedDate, programmeMembershipId &reviewDate...

I had no idea but from a quick search, I *assume* :see_no_evil:, using softDeletedDate is ok and it will still use the index

Reply:
That's definitely worth checking, thanks! 

I had a bit of a look at the EXPAIN output of various indexes, and their performance. It looks like a composite index of softDeleteDate + programmeMembershipId + reviewDate is used in the sort of queries that are run to ID duplicates, but it then doesn't use the traineeId index, so performance ends up being quite poor. Adding traineeId to that composite index makes things a lot better.
My inclination would be to add two (or maybe three) indexes:
1. softDeletedDate - to help the general querying which almost always uses this criterion 
2. composite (traineeId + softDeleteDate + programmeMembershipId + reviewDate) - mainly for duplicate identification
3? composite on the 6 fields used when doing a general assessment search (as in comment above)

As always, additional indexes will slow down data writes, but my assumption is that assessments in general are much more read-heavy. 

What do you guys think?